### PR TITLE
soc: atmel: sam/sam0: Add platform specific initialization

### PR DESCRIPTION
**Pull request by nandojve**
This replace pre kernel initialization by the platform specific initialization call. The platform specific init will configure at very beginning the clocks, flash wait states and cache.

~Supersedes #60425~

### CODE REVIEW

**Review by pdgendt — DISMISSED** [review]

File: soc/arm/atmel_sam0/common/soc_samd5x.c [hunk 1]
```diff
@@ -116,8 +117,14 @@ static int atmel_samd_init(void)
...
-	/* enable the Cortex M Cache Controller */
-	CMCC->CTRL.bit.CEN = 1;
+	/*
+	 * Force Cortex M Cache Controller disabled
+	 *
+	 * It is not clear if regular Cortex-M instructions can be used to
+	 * perform cache maintenance or this is a proprietary cache controller
+	 * that require special SoC support.
+	 */
+	CMCC->CTRL.bit.CEN = 0;
```
Comment (attie-argentum) [inline soc/arm/atmel_sam0/common/soc_samd5x.c:127]: Please break this into a dedicated commit.

File: soc/arm/atmel_sam/sam3x/Kconfig.series [hunk 2]
```diff
@@ -9,9 +9,10 @@ config SOC_SERIES_SAM3X
 	bool "Atmel SAM3X MCU"
 	select ARM
 	select CPU_CORTEX_M3
-	select SOC_FAMILY_SAM
 	select CPU_HAS_ARM_MPU
 	select CPU_CORTEX_M_HAS_DWT
+	select PLATFORM_SPECIFIC_INIT
+	select SOC_FAMILY_SAM
```
Comment (attie-argentum) [inline soc/arm/atmel_sam/sam3x/Kconfig.series:15]: Avoid unnecessary changes? (elsewhere too)

File: soc/arm/atmel_sam/sam3x/soc.c [hunk 3]
```diff
@@ -192,15 +193,7 @@ static ALWAYS_INLINE void clock_init(void)
...
-/**
- * @brief Perform basic hardware initialization at boot.
- *
- * This has to be run at the very beginning thus the init priority is set at
- * 0 (zero).
- *
- * @return 0
- */
-static int atmel_sam3x_init(void)
+void z_arm_platform_init(void)
```
Comment (attie-argentum) [inline soc/arm/atmel_sam/sam3x/soc.c:196]: `z_arm_platform_init()` is a lot earlier than `PRE_KERNEL_1` - are you confident that this will work in all cases?

File: boards/arm/sam_e70_xplained/sam_e70b_xplained_defconfig [hunk 4]
```diff
@@ -6,10 +6,17 @@ CONFIG_SOC_ATMEL_SAME70_EXT_MAINCK=y
...
 CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
 CONFIG_SERIAL=y
 CONFIG_USART_SAM=y
 CONFIG_BOARD_SAM_E70_XPLAINED=y
 CONFIG_WDT_DISABLE_AT_BOOT=y
 CONFIG_BUILD_OUTPUT_HEX=y
+
+# Enable HW stack protection
+CONFIG_HW_STACK_PROTECTION=y
```
[the remark this answers is not in the document]
Comment (nandojve) [reply boards/arm/sam_e70_xplained/sam_e70b_xplained_defconfig]: Done!

File: boards/arm/sam_e70_xplained/sam_e70_xplained_defconfig [hunk 5]
```diff
@@ -6,14 +6,17 @@ CONFIG_SOC_ATMEL_SAME70_EXT_MAINCK=y
 CONFIG_SOC_ATMEL_SAME70_PLLA_MULA=24
 CONFIG_SOC_ATMEL_SAME70_PLLA_DIVA=1
 CONFIG_ARM_MPU=y
+CONFIG_GPIO=y
 CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
 CONFIG_SERIAL=y
 CONFIG_USART_SAM=y
-CONFIG_GPIO=y
```
[the remark this answers is not in the document]
Comment (nandojve) [reply boards/arm/sam_e70_xplained/sam_e70_xplained_defconfig]: I split those changes in other commits. I prefer that usually these configs follow same order to make it easy to compare between variants. I create 3 clusters to group related configs too.

File: soc/arm/atmel_sam0/common/soc_samd5x.c (hunk 1 above)
Comment (nandojve) [reply soc/arm/atmel_sam0/common/soc_samd5x.c:127]: Done!

File: soc/arm/atmel_sam/sam3x/Kconfig.series (hunk 2 above)
Comment (nandojve) [reply soc/arm/atmel_sam/sam3x/Kconfig.series]: ditto

File: soc/arm/atmel_sam/sam3x/soc.c (hunk 3 above)
Comment (nandojve) [reply soc/arm/atmel_sam/sam3x/soc.c:196]: The z_arm_platform_init() seems to be the correct place to configure clocks, cache and any other specific details. This happen at reset which configure SoC ASAP. This speed up core initialization too.

The remaining PRE_KERNEL are initialization that require some kind OS resource.

**Review by pdgendt — APPROVED** [review]

**Review by attie-argentum — APPROVED** [review]
Thanks for the changes!
